flink-staging/flink-streaming/flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/StreamExecutionEnvironment.scala

```diff
@@ -19,396 +19,867 @@
+				function.asInstanceOf[ResultTypeQueryable[T]].getProducedType
+			}
+			else {
+				TypeExtractor.createTypeInfo(classOf[SourceFunction[T]], function.getClass, 0, null, null)
+			}
+
+		val isParallel = function.isInstanceOf[ParallelSourceFunction[T]]
+		val cleanFun = StreamExecutionEnvironment.clean(function)
+		val sourceOperator = new StreamSource[T](cleanFun)
+		new DataStreamSource[T](this.javaEnv, sourceName, typeInfo, sourceOperator,
+			isParallel, sourceName)
+	}
+
+	//	@SuppressWarnings("unchecked")
+	//	private <OUT> DataStreamSource<OUT> addSource(SourceFunction<OUT> function,
+	//		TypeInformation<OUT> typeInfo, String sourceName) {
+	//
+	//			if (typeInfo == null) {
+	//				if (function instanceof GenericSourceFunction) {
+	//					typeInfo = ((GenericSourceFunction<OUT>) function).getType();
+	//				} else {
+	//					typeInfo = TypeExtractor.createTypeInfo(SourceFunction.class,
+	//					function.getClass(), 0, null, null);
+	//				}
+	//			}
+	//
+	//			boolean isParallel = function instanceof ParallelSourceFunction;
+	//
+	//			ClosureCleaner.clean(function, true);
+	//			StreamOperator<OUT, OUT> sourceOperator = new StreamSource<OUT>(function);
+	//
+	//	return new DataStreamSource<OUT>(this, sourceName, typeInfo, sourceOperator,
+	//		isParallel, sourceName);
+	//	}
+
+	/**
+	 * Create a DataStream using a user defined source function for arbitrary
+	 * source functionality.
+	 *
+	 */
+	def addSource[T: ClassTag : TypeInformation](function: Collector[T] => Unit): DataStream[T] = {
+		require(function != null, "Function must not be null.")
+		val sourceFunction = new SourceFunction[T] {
+			val cleanFun = StreamExecutionEnvironment.clean(function)
+
+			override def run(out: Collector[T]) {
+				cleanFun(out)
+			}
+
+			override def cancel() = {}
+		}
+		addSource(sourceFunction)
+	}
+
+	/**
+	 * Triggers the program execution. The environment will execute all parts of
+	 * the program that have resulted in a "sink" operation. Sink operations are
+	 * for example printing results or forwarding them to a message queue.
+	 * <p>
+	 * The program execution will be logged and displayed with a generated
+	 * default name.
+	 *
+	 */
+	def execute() = javaEnv.execute()
+
+	/**
+	 * Triggers the program execution. The environment will execute all parts of
+	 * the program that have resulted in a "sink" operation. Sink operations are
+	 * for example printing results or forwarding them to a message queue.
+	 * <p>
+	 * The program execution will be logged and displayed with the provided name
+	 *
+	 */
+	def execute(jobName: String) = javaEnv.execute(jobName)
+
+	/**
+	 * Creates the plan with which the system will execute the program, and
+	 * returns it as a String using a JSON representation of the execution data
+	 * flow graph. Note that this needs to be called, before the plan is
+	 * executed.
+	 *
+	 */
+	def getExecutionPlan() = javaEnv.getStreamGraph.getStreamingPlanAsJSON
```

Comment:
This change is altering the indentation from spaces to tabs. In scala we are using spaces ;)